### PR TITLE
BUG: Fix Series.groupby raising OutOfBoundsDatetime with DatetimeIndex and month name

### LINKED ISSUE #48509: BUG: groupby() attempts to converts name of Series if name is a month and index is DateTime
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd

s = pd.Series([1], index=pd.date_range('2022-01-01', periods=1), name='JAN')
res = s.groupby(s).cumcount()
print(res)
```

### Issue Description

Doing groupby() on a Series, with a DatetimeIndex and a name that is a month abbreviation
triggers an attempt to convert the name into a DateTime, which fails/throws.

Either using integer index or having a non month name avoids the issue.

It was reproduced on both MacOS and Windows.

(The issue showed up in a DataFrame.apply(), where the column name happened to be a month name)

### Expected Behavior

It is expected that the column name is left as is.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : e8093ba372f9adfe79439d90fe74b0b5b6dea9d6
python           : 3.10.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.6.0
Version          : Darwin Kernel Version 20.6.0: Tue Jun 21 20:50:28 PDT 2022; root:xnu-7195.141.32~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 1.4.3
numpy            : 1.23.1
pytz             : 2022.1
dateutil         : 2.8.2
setuptools       : 63.4.1
pip              : 22.1.2
Cython           : None
pytest           : 7.1.2
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.8.6
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : 1.3.5
brotli           : 
fastparquet      : None
fsspec           : None
gcsfs            : None
markupsafe       : None
matplotlib       : 3.5.2
numba            : None
numexpr          : 2.8.3
odfpy            : None
openpyxl         : 3.0.10
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : 1.4.39
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by yuanx749**
Fix the bug as discussed in #48509 @topper-123 
- [x] closes #48509
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -978,11 +979,13 @@ def is_in_obj(gpr) -> bool:
             return False
         try:
             return gpr is obj[gpr.name]
-        except (KeyError, IndexError, InvalidIndexError):
```
Comment (rhshadrach) [inline pandas/core/groupby/grouper.py:981]: It seems like any error we hit here will want to be ignored; what are thoughts on just doing `try: ... except Exception: return False`?

cc @jbrockmendel @mroeschke

Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:981]: How would you get an arbitrary exception here?  A couple years back we made a concerted effort to be more specific in what we caught

Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:981]: I think there are other issues caused by not being strict _enough_ here e.g. #51818

Comment (rhshadrach) [reply pandas/core/groupby/grouper.py:981]: I don't think you can; what I'm think is that any exception that can possibly be hit here we want to ignore. I'm not certain that's true though.

Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:981]: I think the listed exceptions here are ones that make sense if `gpr.name` is not found in `obj.columns`.  Any other exceptions mean Something Has Gone Wrong and we would want a bug report to be filed

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @yuanx749

**Pull request merged**